pkg/lib/hooks.js

```diff
@@ -0,0 +1,291 @@
+ * very same object, and the effect hook is only run once.  If we
+ * would use "options" directly as a dependency of the effect hook,
+ * the channel would be closed and opened on every render. This is
+ * very inefficient, doesn't give the asynchronous channel time to do
+ * its job, and will also lead to infinite loops when events on the
+ * channel cause re-renders (which in turn will run the effect hook
+ * again, which will cause a new event, ...).
+ */
+
+export function useDeepEqualMemo(value) {
+    const ref = useRef(value);
+    if (!deep_equal(ref.current, value))
+        ref.current = value;
+    return ref.current;
+}
+
+/* - useFile(path, options)
+ *
+ * function Component() {
+ *   const content = useFile("/etc/hostname", { superuser: "try" });
+ *   const [content, error] = useFileWithError("/etc/hostname", { superuser: "try" });
+ *
+ *   ...
+ * }
+ *
+ * The "path" and "options" parameters are passed unchanged to
+ * cockpit.file().  Thus, if you need to parse the content of the
+ * file, the best way to do that is via the "syntax" option.
+ *
+ * The "content" variable will reflect the content of the file
+ * "/etc/hostname". When the file changes on disk, the component will
+ * be re-rendered with the new content.
+ *
+ * When the file does not exist or there has been some error reading
+ * it, "content" will be false.
+ *
+ * The "error" variable will contain any errors encountered while
+ * reading the file.  It is false when there are no errors.
+ *
+ * When the file does not exist, "error" will be false.
+ *
+ * The "content" and "error" variables will be null until the file has
+ * been read for the first time.
+ *
+ * useFile and useFileWithError are pretty much the same. useFile will
+ * hide the exact error from the caller, which makes it slightly
+ * cleaner to use when the exact error is not part of the UI. In the
+ * case of error, useFile will log that error to the console.
+ */
+
+export function useFileWithError(path, options, hook_options) {
+    const [content_and_error, setContentAndError] = useState([null, null]);
+    const memo_options = useDeepEqualMemo(options);
+    const memo_hook_options = useDeepEqualMemo(hook_options);
+
+    useEffect(() => {
+        const handle = cockpit.file(path, memo_options);
+        handle.watch((data, tag, error) => {
+            setContentAndError([data || false, error || false]);
+            if (!data && memo_hook_options && memo_hook_options.log_errors)
+                console.warn("Can't read " + path + ": " + (error ? error.toString() : "not found"));
+        });
+        return function () {
+            handle.close();
+        };
+    }, [path, memo_options, memo_hook_options]);
+
+    return content_and_error;
+}
+
+export function useFile(path, options) {
+    const [content] = useFileWithError(path, options, { log_errors: true });
+    return content;
+}
+
+/* - useObject(create, destroy, deps)
```

Comment:
Missing the "comps" argument. Also, as this is a library, could you spell out the argument names? It's not clear what  "comps" does, looks like a list of comparison functions? Could you please document that?

Reply:
Yes, done.